### PR TITLE
AF-2766: Dashbuilder Prometheus data-set provider

### PR DESCRIPTION
**Pull request by jesuino**
The dataset provider impl for Prometheus.

Part of an ensemble. Merge with:

https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/1566
https://github.com/kiegroup/appformer/pull/1091

### CODE REVIEW

File: kie-soup-dataset/kie-soup-dataset-api/src/main/java/org/dashbuilder/dataset/impl/PrometheusDataSetDefBuilderImpl.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import org.dashbuilder.dataset.def.DataSetDef;
+import org.dashbuilder.dataset.def.PrometheusDataSetDef;
+import org.dashbuilder.dataset.def.PrometheusDataSetDefBuilder;
+
+public class PrometheusDataSetDefBuilderImpl extends AbstractDataSetDefBuilder<PrometheusDataSetDefBuilderImpl> implements PrometheusDataSetDefBuilder<PrometheusDataSetDefBuilderImpl> {
+
+    @Override
+    public PrometheusDataSetDefBuilderImpl serverUrl(String serverUrl) {
+        ((PrometheusDataSetDef) def).setServerUrl(serverUrl);
```
Comment (caponetto) [inline kie-soup-dataset/kie-soup-dataset-api/src/main/java/org/dashbuilder/dataset/impl/PrometheusDataSetDefBuilderImpl.java:26]:
```suggestion
        ((PrometheusDataSetDef) super.def).setServerUrl(serverUrl);
```
Consider adding a `super` for this and the other method below.
IMO it improves readability.

File: kie-soup-dataset/kie-soup-dataset-prometheus/src/main/java/org/dashbuilder/dataprovider/prometheus/PrometheusDataSetProvider.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                int i = 0;
+                for (String key : metricColumns) {
+                    row[i++] = metric.get(key);
+                }
+                row[i++] = value.getTimestamp();
+                row[i] = value.getValue();
+                dataSet.addValuesAt(dataSet.getRowCount(), row);
+            }
+        }
+        return dataSet;
```
Comment (caponetto) [inline kie-soup-dataset/kie-soup-dataset-prometheus/src/main/java/org/dashbuilder/dataprovider/prometheus/PrometheusDataSetProvider.java:130]: I don't like 3 nested `for` loops but I guess there is no other way, right?

File: kie-soup-dataset/kie-soup-dataset-api/src/main/java/org/dashbuilder/dataset/def/PrometheusDataSetDef.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    @Size(min = 1, groups = {PrometheusDataSetDefValidation.class})
+    @NotNull(groups = {PrometheusDataSetDefValidation.class})
+    protected String query;
+
+    private String user;
+    private String password;
+
+    public PrometheusDataSetDef() {
+        super.setProvider(DataSetProviderType.PROMETHEUS);
+        serverUrl = "http://localhost:9090";
```
Comment (caponetto) [inline kie-soup-dataset/kie-soup-dataset-api/src/main/java/org/dashbuilder/dataset/def/PrometheusDataSetDef.java:39]: Should this port be a fixed value?

File: kie-soup-dataset/kie-soup-dataset-prometheus/src/main/java/org/dashbuilder/dataprovider/prometheus/PrometheusDataSetProvider.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+public class PrometheusDataSetProvider implements DataSetProvider, DataSetDefRegistryListener {
+
+    public static final String VALUE_COLUMN = "VALUE";
+    public static final String TIME_COLUMN = "TIME";
+
+    protected StaticDataSetProvider staticDataSetProvider;
+    protected Logger log = LoggerFactory.getLogger(PrometheusDataSetProvider.class);
+
+    private static PrometheusDataSetProvider SINGLETON = null;
```
Comment (caponetto) [inline kie-soup-dataset/kie-soup-dataset-prometheus/src/main/java/org/dashbuilder/dataprovider/prometheus/PrometheusDataSetProvider.java:53]:
```suggestion
    private static PrometheusDataSetProvider instance = null;
```

File: kie-soup-dataset/kie-soup-dataset-prometheus/src/main/java/org/dashbuilder/dataprovider/prometheus/client/PrometheusClient.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class PrometheusClient {
+
+    private static final String DEFAULT_BASE_URL = "http://localhost:9090";
```
Comment (caponetto) [inline kie-soup-dataset/kie-soup-dataset-prometheus/src/main/java/org/dashbuilder/dataprovider/prometheus/client/PrometheusClient.java:27]: Should this port be a fixed value?

File: kie-soup-dataset/kie-soup-dataset-api/src/main/java/org/dashbuilder/dataset/json/PrometheusDefJSONMarshaller.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.dashbuilder.dataset.def.PrometheusDataSetDef;
+import org.dashbuilder.json.JsonObject;
+
+import static org.dashbuilder.dataset.json.DataSetDefJSONMarshaller.ALL_COLUMNS;
+import static org.dashbuilder.dataset.json.DataSetDefJSONMarshaller.isBlank;
+
+public class PrometheusDefJSONMarshaller implements DataSetDefJSONMarshallerExt<PrometheusDataSetDef> {
+
+    public static PrometheusDefJSONMarshaller INSTANCE = new PrometheusDefJSONMarshaller();
```
[the remark this answers is not in the document]
Comment (jesuino) [reply kie-soup-dataset/kie-soup-dataset-api/src/main/java/org/dashbuilder/dataset/json/PrometheusDefJSONMarshaller.java:26]: changed this field to be final as well, so not changing the case

File: kie-soup-dataset/kie-soup-dataset-prometheus/src/main/java/org/dashbuilder/dataprovider/prometheus/PrometheusDataSetProvider.java (hunk 2 above)
Comment (jesuino) [reply kie-soup-dataset/kie-soup-dataset-prometheus/src/main/java/org/dashbuilder/dataprovider/prometheus/PrometheusDataSetProvider.java:130]: Unfortunately no, for each metric column we add a dataset column, hence why we have the loop from line 122 - but it is expected to have less than 10 records as metric columns

File: kie-soup-dataset/kie-soup-dataset-api/src/main/java/org/dashbuilder/dataset/def/PrometheusDataSetDef.java (hunk 3 above)
Comment (jesuino) [reply kie-soup-dataset/kie-soup-dataset-api/src/main/java/org/dashbuilder/dataset/def/PrometheusDataSetDef.java:39]: This value should be overriden later by actual values - I set the default as the default Prometheus port, but usually it is overriden

**Comment by jesuino** [comment]
Hello @caponetto  - thanks for the feedback! I just made the requested changes, please let me know if you have any question

**Review by caponetto — APPROVED** [review]

**Review by ederign — APPROVED** [review]

**Pull request merged**